index.js

```diff
@@ -1,35 +1,28 @@
 // Public intefrace
 module.exports = promisedPipe
 
+const chain = (q, fn) => q.then(fn)
+
 function promisedPipe(...fns) {
     if (fns.length < 1) {
         throw Error('pipe requires at least one argument')
     }
 
-    fns.forEach((fn, i) => {
-        if (typeof fn !== 'function') {
-            throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fn}"`)
+    let i = -1
+    while (++i < fns.length) {
+        if (typeof fns[i] !== 'function') {
+            throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fns[i]}"`)
         }
-    })
+    }
 
-    /*
-     * Does the same as q.promised
-     * https://github.com/kriskowal/q/wiki/API-Reference#qpromisedfunc
-     *
-     * Creates a new version of func that accepts any combination of promise and non-promise
-     * values, converting them to their fulfillment values before calling the original func.
-     * The returned version also always returns a promise: if func does a return or throw,
-     * then Q.promised(func) will return fulfilled or rejected promise, respectively.
-     *
-     */
-    const promised = fn => (...args) => Promise.all(args).then(_args => fn.apply(null, _args))
+    // shift out the 1st function for multiple arguments
+    const start = fns.shift()
 
-    /*
-     * Does the same as Ramda.pipe
-     * http://ramdajs.com/docs/#pipe
-     *
-     * Performs left-to-right function composition. The leftmost function may have any arity;
-     * the remaining functions must be unary.
-     */
-    return (...args) => fns.map(promised).reduce((acc, cur) => [cur.apply(null, acc)], args)[0]
+    return (...args) => {
+        try {
+            return fns.reduce(chain, Promise.resolve(start(...args)))
```

Comment:
@kigiri looks better now, thanks!

One more thing. Since you started the performance conversation, wouldn't it harm the performance when you put `try / catch` around the whole `fns.reduce` function?
AFAIK `try / catch` prevents V8 from doing a lot of optimizations and it might be smarter to only affect `start(...args)`.